contracts/src/token/erc20/extensions/permit.rs

```diff
@@ -0,0 +1,196 @@
+//!
+//! Adds the `permit` method, which can be used to change an account’s
+//! ERC20 allowance (see [`crate::token::erc20::IErc20::allowance`])
+//! by presenting a message signed by the account.
+//! By not relying on [`crate::token::erc20::IErc20::approve`],
+//! the token holder account doesn’t need to send a transaction,
+//! and thus is not required to hold Ether at all.
+use alloy_primitives::{b256, keccak256, Address, B256, U256};
+use alloy_sol_types::{sol, SolType};
+use stylus_proc::{external, sol_storage, SolidityError};
+use stylus_sdk::{block, prelude::StorageType, storage::TopLevelStorage};
+
+use crate::{
+    token::erc20::IErc20Internal,
+    utils::{
+        cryptography::{ecdsa, eip712::IEIP712},
+        nonces::Nonces,
+    },
+};
+
+const PERMIT_TYPEHASH: B256 =
+    b256!("6e71edae12b1b97f4d1f60370fef10105fa2faae0126114a169c64845d6126c9");
+
+type StructHashTuple = sol! {
+    tuple(bytes32, address, address, uint256, uint256, uint256)
+};
+
+/// Calculates `keccak256` hash value for a struct used in Permit extension.
+///
+/// # Arguments
+///
+/// * `owner` - Account that owns the tokens.
+/// * `spender` - Account that will spend the tokens.
+/// * `value` - The number of tokens being permitted to transfer by `spender`.
+/// * `deadline` - Deadline for the `permit` action.
+pub fn prepare_struct_hash(
+    owner: Address,
+    spender: Address,
+    value: U256,
+    nonce: U256,
+    deadline: U256,
+) -> B256 {
+    keccak256(StructHashTuple::encode_params(&(
+        *PERMIT_TYPEHASH,
+        owner,
+        spender,
+        value,
+        nonce,
+        deadline,
+    )))
+}
+
+sol! {
+    /// Indicates an error related to the fact that
+    /// permit deadline has expired.
+    #[derive(Debug)]
+    #[allow(missing_docs)]
+    error ERC2612ExpiredSignature(uint256 deadline);
+
+    /// Indicates an error related to the issue about mismatched signature.
+    #[derive(Debug)]
+    #[allow(missing_docs)]
+    error ERC2612InvalidSigner(address signer, address owner);
+}
+
+/// A Permit error.
+#[derive(SolidityError, Debug)]
+pub enum Error {
+    /// Indicates an error related to the fact that
+    /// permit deadline has expired.
+    ExpiredSignature(ERC2612ExpiredSignature),
+    /// Indicates an error related to the issue about mismatched signature.
+    InvalidSigner(ERC2612InvalidSigner),
+}
+
+sol_storage! {
+    /// State of a Permit Contract.
+    pub struct Permit<T: IEIP712 + StorageType>{
+        /// Nonces contract.
+        Nonces nonces;
+        /// EIP-712 contract. Must implement [`IEIP712`] trait.
+        T eip712;
+    }
+}
+
+/// NOTE: Implementation of [`TopLevelStorage`] to be able use `&mut self` when
+/// calling other contracts and not `&mut (impl TopLevelStorage +
+/// BorrowMut<Self>)`. Should be fixed in the future by the Stylus team.
+unsafe impl<T: IEIP712 + StorageType> TopLevelStorage for Permit<T> {}
+
+#[external]
+impl<T: IEIP712 + StorageType> Permit<T> {
```

Comment:
Why do we require a `StorageType` bound here?

Reply:
```
error[E0277]: the trait bound `T: stylus_sdk::prelude::StorageType` is not satisfied
  --> contracts/src/token/erc20/extensions/permit.rs:57:1
   |
57 | / sol_storage! {
58 | |     /// State of a Permit Contract.
59 | |     pub struct Erc20Permit<T: IEip712>{
60 | |         /// ERC-20 contract.
...  |
68 | |     }
69 | | }
   | |_^ the trait `stylus_sdk::prelude::StorageType` is not implemented for `T`
```